subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1065 @@
+		memcpy(&data_msg_rx->block, block, sizeof(struct audio_data));
+		data_msg_rx->tx_handle = tx_handle;
+		data_msg_rx->response_cb = data_in_response_cb;
+
+		ret = data_fifo_block_lock(rx_handle->thread.msg_rx, (void **)&data_msg_rx,
+					   sizeof(struct audio_module_message));
+		if (ret) {
+			data_fifo_block_free(rx_handle->thread.msg_rx, (void **)&data_msg_rx);
+
+			LOG_DBG("Module %s failed to queue block, ret %d", rx_handle->name, ret);
+			return ret;
+		}
+
+		LOG_DBG("Block sent to module %s", rx_handle->name);
+
+	} else {
+		LOG_DBG("Receiving module %s is in an invalid state %d", rx_handle->name,
+			rx_handle->state);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Send to modules TX message queue.
+ *
+ * @param handle  The handle for this modules instance
+ * @param block   A pointer to the audio block
+ *
+ * @return 0 if successful, error value
+ */
+static int send_to_tx_fifo(struct audio_module_handle *handle, struct audio_data *block)
+{
+	int ret;
+	struct audio_module_message *data_msg_tx;
+
+	/* Take a TX message */
+	ret = data_fifo_pointer_first_vacant_get(handle->thread.msg_tx, (void **)&data_msg_tx,
+						 K_FOREVER);
+	if (ret) {
+		LOG_DBG("No free message for module %s, ret %d", handle->name, ret);
+		return ret;
+	}
+
+	/* Configure audio block */
+	memcpy(&data_msg_tx->block, block, sizeof(struct audio_data));
+	data_msg_tx->tx_handle = handle;
+	data_msg_tx->response_cb = block_release_cb;
+
+	/* Send audio block to modules output message queue */
+	ret = data_fifo_block_lock(handle->thread.msg_tx, (void **)&data_msg_tx,
+				   sizeof(struct audio_module_message));
+	if (ret) {
+		LOG_DBG("Failed to send block to output of module %s, ret %d", handle->name, ret);
+
+		data_fifo_block_free(handle->thread.msg_tx, (void **)&data_msg_tx);
+
+		k_sem_take(&handle->sem, K_NO_WAIT);
+
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Send output audio block to next module(s)
+ *
+ * @param handle  The handle for this modules instance
+ * @param block   A pointer to the audio block
+ *
+ * @return 0 if successful, error value
+ */
+static int send_to_connected(struct audio_module_handle *handle, struct audio_data *block)
+{
+	int ret;
+	struct audio_module_handle *handle_to;
+
+	if (handle->dest_count == 0) {
+		LOG_DBG("No where to send the data from module %s so releasing it", handle->name);
+
+		k_mem_slab_free(handle->thread.data_slab, (void **)block->data);
+
+		return 0;
+	}
+
+	k_mutex_lock(&handle->dest_mutex, K_FOREVER);
+
+	k_sem_init(&handle->sem, handle->dest_count, handle->dest_count);
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&handle->hdl_dest_list, handle_to, node) {
+		ret = data_tx(handle, handle_to, block, &block_release_cb);
+		if (ret) {
+			LOG_DBG("Failed to send to module %s from %s, ret %d", handle_to->name,
+				handle->name, ret);
+			k_sem_take(&handle->sem, K_NO_WAIT);
+		}
+	}
+
+	if (handle->data_tx && handle->thread.msg_tx) {
+		ret = send_to_tx_fifo(handle, block);
```

Comment:
The data_tx sends data to connected module's RX queue, whereas the send_to_tx_fifo sends data to the TX queue of the module. This can then be picked up by software outside of the audio system.

Reply:
Can we then do a rename to make this clearer?